### PR TITLE
dynamodocstore: query plan with indexes

### PR DESCRIPTION
**Pull request by jba**
Take table indexes into account when deciding how to run a query.

Add tests for query planning.

Updates #1832. This PR builds and runs the RPCs, but we also need to
report the plan to the user.

### CODE REVIEW

File: internal/docstore/dynamodocstore/query.go [hunk 1]
```diff
@@ -47,7 +58,7 @@ func (c *collection) RunGetQuery(ctx context.Context, q *driver.Query) (driver.D
 	return it, nil
 }
 
-func (c *collection) planQuery(q *driver.Query) (*queryRunner, error) {
+func (c *collection) planQueryOld(q *driver.Query) (*queryRunner, error) {
```
Comment (shantuo) [inline internal/docstore/dynamodocstore/query.go]: Why keeps it?

File: internal/docstore/dynamodocstore/query_test.go [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+			},
+		},
+		{
+			desc: "equality filter on table partition, filter on global index sort, bad projection",
+			// Although there is a global index that matches the filters best, it doesn't
+			// have the necessary fields. So we query against the table.
+			globalIndexPartitionKey: "tableP",
+			globalIndexSortKey:      "globalS",
+			globalIndexFields:       []string{"other"},
+			query: &driver.Query{Filters: []driver.Filter{
```
Comment (shantuo) [inline internal/docstore/dynamodocstore/query_test.go]: Consider adding a case that the query asks for some paths that the global index doesn't contain.

File: internal/docstore/dynamodocstore/query.go (hunk 1 above)
Comment (jba) [reply internal/docstore/dynamodocstore/query.go]: Fixed.

File: internal/docstore/dynamodocstore/query_test.go (hunk 2 above)
Comment (jba) [reply internal/docstore/dynamodocstore/query_test.go]: Thanks, this case actually does that but not in a clear way. I fixed it.

**Pull request merged**